core-rust/core-api-server/src/core_api/conversions/substates/access_controller.rs

```diff
@@ -12,11 +13,95 @@ pub fn to_api_access_controller_substate(
     Ok(field_substate_versioned!(
         substate,
         AccessControllerFieldState,
-        substate => {
-            let data = scrypto_encode(substate).unwrap();
+        AccessControllerSubstate {
+            controlled_asset,
+            timed_recovery_delay_in_minutes,
+            recovery_badge,
+            state: (
+                primary_role_locking_state,
+                primary_role_recovery_attempt_state,
+                primary_role_badge_withdraw_attempt_state,
+                recovery_role_recovery_attempt_state,
+                recovery_role_badge_withdraw_attempt_state,
+            )
         },
         Value {
-            data_struct: Box::new(to_api_data_struct_from_bytes(context, &data)?),
+            controlled_vault_entity: Box::new(to_api_entity_reference(
+                context,
+                controlled_asset.as_node_id()
+            )?),
+            timed_recovery_delay_minutes: timed_recovery_delay_in_minutes
+                .as_ref()
+                .map(|minutes| to_api_u32_as_i64(*minutes)),
+            recovery_badge_resource_address: to_api_resource_address(context, recovery_badge)?,
+            is_primary_role_locked: matches!(
```

Comment:
Minor, but would slightly rather do an explicit match to ensure we get a compile error if new variants get added.